### ISSUE TITLE
when can we update to distil or faster whisper?

### ISSUE THREAD

**Issue by striker9k** [rpdrewes/whisper-websocket-server#4]
Hey! I noticed we're still running whisper.ccp, i think it would be brillant if we can upgrade to either distill-whisper or faster-whisper to get the best out of it. Thanks!

**Comment by rpdrewes** [comment]
I see that recently ggerganov's project, whose code I find very easy to work with, has some integration with distil-whisper. I am looking into integrating with that since it promises big performance improvement.

It would be helpful to me to know that you are using whisper-websocket-server!

**Comment by rpdrewes** [comment]
I've looked into this and have the following observations:

1) The latest whisper.cpp (1.5.1) includes the distil medium.en model by default. It does run faster than the old non-distil medium model in my tests (about 2.5 times faster on my CPU). All you need to do to get whisper-websocket-server to use the distil model is to change the model from tiny to the medium distil model in the "transcmd" in minimal.py. Super simple.

2) However, in my tests, distil-medium still runs much slower than the "tiny" model that this project uses by default, though I'm sure distil-medium has much better accuracy. There is a distil-small model available for download from OpenAI, and that might run as fast as the regular (non distil) tiny model, but I have had some problems with the distil-small model. It does not seem to be recognized as a distil model at load time by whisper.cpp, and on a simple test it does not even transcribe the full input. So maybe something is wrong with it?

In short I don't see a simple way right now to increase performance of whisper-websocket-server by switching to a distil model due to unavailability of working small or tiny distil models. I might look into faster-whisper, and so could you.

I will point out also that the current invocation of the whisper.cpp from whisper-websocket-server is super simple and inefficient (it loads the model new on each recognition!) so there is lots of room to reduce latency there. But the whole thing still works great for me as it is!

Rich

**Comment by rpdrewes** [comment]
I have had some great results integrating with faster-whisper. I also improved the very simplistic earlier implementation in minimal.py that used a temporary file to instead do in-memory buffering of sound data. This gives better performance and allows multiple requests to be in progress at once (not tested extensively). Also, using the higher-performance faster-whisper allows using using a better recognition model (like "base.en") with same or better performance compared to what was used before ("tiny.en") with whisper.cpp. And this is done entirely in Python faster-whisper without using system() to invoke whisper.cpp. I will publish these changes to the git repo within a few days.

I still use this DAILY to provide voice recognition services for my de-googled Android phone (LineageOS) and it is FANTASTIC. It seems to outperform Apple and Google provided voice recognition services. Does anyone else do this? If not, WHY NOT???

**Comment by rpdrewes** [comment]
I have updated repo to reflect a change from whisper.cpp to faster-whisper. This is faster even with a larger model (base.en vs tiny.en) and other improvements are also incorporated, including: audio data is stored in memory rather than temporary file so it is faster and supports multiple connections at once, and there is no system() to invoke whisper.cpp so it is faster and more secure, and there are instructions for use with docker for security isolation and simplicity of management.

**Comment by striker9k** [comment]
I love you Rich. Thanks for all the effort you put in, it works flawlessly. (and apologies for not contributing, unfortunately, I am no dev :/  )
